File: apps/backend/scout/clients/geocoding/stub.py
"""Offline stub geocoder — deterministic hits for tests and stubbed dev."""

from __future__ import annotations

from scout.clients.geocoding.protocol import AddressHit, GeocodingProvider

_STUB_HITS: tuple[AddressHit, ...] = (
    AddressHit(
        id="stub-14th-u",
        label="1400 U Street Northwest, Washington, DC 20009, United States",
        lon=-77.0366,
        lat=38.9169,
    ),
    AddressHit(
        id="stub-dupont",
        label="Dupont Circle, Washington, District of Columbia 20009, United States",
        lon=-77.0369,
        lat=38.9097,
    ),
    AddressHit(
        # Named-place fixture (DEC-026) — exercises the same `AddressHit`
        # shape a real `dc_points_of_interest` hit would produce.
        id="stub-poi-national-building-museum",
        label="National Building Museum, 401 F St NW, Washington, DC 20001",
        lon=-77.0157,
        lat=38.8983,
    ),
)


class StubGeocodingProvider(GeocodingProvider):
    async def search(self, query: str, *, limit: int = 5) -> list[AddressHit]:
        trimmed = query.strip().lower()
        if not trimmed:
            return []
        if "national building" in trimmed or "building museum" in trimmed:
            return [_STUB_HITS[2]] if limit >= 1 else []
        if "dupont" in trimmed or "14th" in trimmed or "14 " in trimmed:
            return list(_STUB_HITS[:2][:limit])
        return [_STUB_HITS[0]] if limit >= 1 else []

    async def reverse(self, lon: float, lat: float) -> AddressHit:
        # Deliberately a non-address label so a developer who clicks
        # "Use my location" against this stub can't mistake it for a
        # real reverse-geocode result.
        return AddressHit(
            id="stub-reverse",
            label=(
                "[STUB] Reverse-geocode fixture — "
                "override SCOUT_GEOCODING_PROVIDER for a live-backed dev stack"
            ),
            lon=lon,
            lat=lat,
        )

File: apps/web/lib/providers/geocoding/stub.ts
import type { AddressHit, GeocodingProvider, GeocodingSearchOptions } from "./protocol";

/** Deterministic hits for Vitest / Playwright (no live geocoder upstream). */
export const STUB_SEARCH_HITS: readonly AddressHit[] = [
  {
    id: "stub-14th-u",
    label: "1400 U Street Northwest, Washington, DC 20009, United States",
    lon: -77.0366,
    lat: 38.9169,
  },
  {
    id: "stub-dupont",
    label: "Dupont Circle, Washington, District of Columbia 20009, United States",
    lon: -77.0369,
    lat: 38.9097,
  },
];

function sleep(signal: AbortSignal | undefined): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted === true) {
      reject(signal.reason);
      return;
    }

    const onAbort = (): void => {
      signal?.removeEventListener("abort", onAbort);
      reject(signal?.reason ?? new DOMException("Aborted", "AbortError"));
    };

    signal?.addEventListener("abort", onAbort);
    queueMicrotask(() => {
      signal?.removeEventListener("abort", onAbort);
      if (signal?.aborted === true) {
        reject(signal.reason);
        return;
      }
      resolve();
    });
  });
}

async function filteredSearchHits(
  query: string,
  opts: GeocodingSearchOptions | undefined,
): Promise<readonly AddressHit[]> {
  const trimmed = query.trim().toLowerCase();
  const limit = typeof opts?.limit === "number" ? Math.min(opts.limit, 10) : 5;

  if (trimmed.length < 3) {
    return [];
  }

  const firstHit = STUB_SEARCH_HITS[0];
  const matches: readonly AddressHit[] =
    trimmed.includes("dupont") || trimmed.includes("14th") || trimmed.includes("14 ")
      ? STUB_SEARCH_HITS
      : firstHit
        ? [firstHit]
        : [];

  return matches.slice(0, limit);
}

export const stubGeocodingProvider: GeocodingProvider = {
  async search(query: string, options: GeocodingSearchOptions | undefined, signal?) {
    await sleep(signal);
    return [...(await filteredSearchHits(query, options))];
  },

  async reverse(_lon: number, _lat: number, signal?) {
    await sleep(signal);
    void _lon;
    void _lat;
    // Label is intentionally an obvious non-address so a developer who clicks
    // "Use my location" cannot mistake this fixture for a real reverse-geocode
    // result. Flip NEXT_PUBLIC_SCOUT_GEOCODING_PROVIDER=backend (or run
    // `make docker-up-realistic-run`) to hit Scout's `/api/geocode/reverse`
    // (Photon upstream via the backend adapter, per DEC-022) instead.
    return {
      id: "stub-reverse-1",
      label:
        "[STUB] Reverse-geocode fixture — set NEXT_PUBLIC_SCOUT_GEOCODING_PROVIDER=backend for real results",
      lon: -77.0324,
      lat: 38.8997,
    };
  },
};
